test/cli/test_cli.ml

```diff
@@ -8,19 +8,32 @@ let pp_status ppf = function
   | WSTOPPED _ -> Format.fprintf ppf "WSTOPPED _"
 
 let main ocamlformat =
-  let test ?(stdin = "") ~name ~args () =
-    let cmd = Printf.sprintf "%s %s 2>&1" ocamlformat args in
-    let proc = Unix.open_process cmd in
-    let proc_out, proc_in = proc in
+  let ocamlformat =
+    (* Turn into absolute path *)
+    if Filename.is_relative ocamlformat then
+      Filename.concat (Unix.getcwd ()) ocamlformat
+    else
+      ocamlformat
+  in
+  let test ?(env = [||]) ?pwd ?(stdin = "") ~name ~args () =
+    let pp_cd pp = function
+      | Some d -> Format.fprintf pp "cd '%s'; " d
+      | None -> ()
+    in
+    let cmd = Format.asprintf "%a%s %s 2>&1" pp_cd pwd ocamlformat args in
+    let proc = Unix.open_process_full cmd env in
+    let proc_out, proc_in, _ = proc in
     Stdio.Out_channel.output_string proc_in stdin ;
     Stdio.Out_channel.close proc_in ;
     let out = Stdio.In_channel.input_all proc_out in
-    let stat = Unix.close_process proc in
+    let stat = Unix.close_process_full proc in
     Format.printf "%aargs: %s\n%s%a\n\n" pp_banner name args out pp_status stat
   in
   let impl_file = Stdio.In_channel.read_all "sample/b.ml" in
   let intf_file = Stdio.In_channel.read_all "sample/a.mli" in
   let syntax_error_file = Stdio.In_channel.read_all "sample/syntax_error.ml" in
+
+  (* CLI actions *)
```

Comment:
I think extracting functions will make this clearer than a comment: `test_cli_actions (); test_user_errors()` etc.

Reply:
Yes this would help.